### DISCUSSION TITLE
Add `fromEventTarget` method, to create events from side events in the unified way

### DISCUSSION THREAD

**Discussion by sergeysova** [zerobias/effector#490] [Ideas]
What is it allows?
- Subscribe to browser events ([websocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/message_event), [DOM elements](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener), [service workers](https://developer.mozilla.org/en-US/docs/Web/API/NotificationEvent), [match media](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList))
- Subscribe to nodejs [events](https://nodejs.org/api/events.html#events_emitter_addlistener_eventname_listener)
- Subscribe to side event emitters ([socket.io](https://socket.io/docs/v4/listening-to-events/), [eventemitter3](https://www.npmjs.com/package/eventemitter3))

How:

Add method `fromEventTarget` that receives EventTarget/EventEmitter, event name, and returns [`Event`](https://effector.dev/docs/api/effector/event/).

NodeJS:

```ts
const http2 = require('http2');
const client = http2.connect('https://localhost:8443', options);
const req = client.request({ ':path': '/' })

const responseReceived = fromEventTarget(client, 'response');
const dataReceived = fromEventTarget(client, 'data')
```

WebSocket:

```ts
const socket = new WebSocket('ws://localhost:8080');
const messageReceived = fromEventTarget(socket, 'message');
```

socket.io-client:

```ts
import { io } from 'socket.io-client';
const socket = io('https://server-domain.com')
const details = fromEventTarget(socket, 'details');
```

ServiceWorker:

```ts
const notificationClicked = fromEventTarget(self, 'notificationclick');
```

Window:

```ts
const mouseMoved = fromEventTarget(window, "mousemove");
```

MediaQueryList:

```ts
const mql = window.matchMedia('(max-width: 600px)');
const matchChanged = fromEventTarget(mql, 'changed');
```

eventemitter3:

```ts
const EE = require('eventemitter3');
const emitter = new EE();
const itIsDone = fromEventTarget(EE, 'done');

emitter.emit('done', value);
```

---

Questions:

### 1. What if event listener should receive multiple parameters?

Maybe add `mapParams`

```ts
const effectorEvent = fromEventTarget(target, 'event_name', {
  mapParams: (first, second, third) => ({ first, second, third }),
})
effectorEvent.watch(({ first, second, third }) => {})
```

### 2. [EventTarget.addEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) has options, how can I pass it?

Add `options`:

```ts
const effectorEvent = fromEventTarget(element, 'click', { capture: true, once: true })
```

**Comment by vlakam** [comment]
How would you do removeEventListener?

**Reply by sergeysova** [reply]
If you want to remove the listener, you should not use this method. It is just for static event creation

**Comment by sergeysova** [comment]
Naive implementation sample:

```ts
export function fromEventTarget(target, eventName, { mapParams: (a) => a, options } = {}) {
  const event = createEvent();
  const handler = (...all) => event(mapParams(...all))
  if (target.addListener) target.addListener(eventName, handler, options);
  else if (target.addEventListener) target.addEventListener(eventName, handler, op);
  else if (target.on) target.on(eventName, handler);
  else {
    throw new TypeError('target passed to `fromEventTarget` should be an EventTarget or EventEmitter instance')
  }
  return event
}
```

**Reply by rjkz808** [reply]
The same API is already implemented in [rxjs](https://rxjs.dev/api/index/function/fromEvent), I think their interfaces could be reused in effector as well.

**Comment by meff34** [comment]
Do we *really* need to repeat rxjs? Why not let the user take care of the subscription himself?  That seems to be a bit more universal solution.

```typescript
type Unsubscribe = () => void
type Trigger<T> = (x: T) => void
type Subscriber<T> = (trigger: Trigger<T>) => Unsubscribe

declare const fromContinuousCallback: <T>(subscriber: Subscriber<T>) => [
  resultEvent: Event<T>,
  unsubscribe: Event<void>
]
```

nodejs:
```javascript
const http2 = require('http2');
const client = http2.connect('https://localhost:8443', options);
const req = client.request({ ':path': '/' })

const [responseReceived, noop] = fromContinuousCallback(trigger => {
  client.on('response', trigger)
});
const [dataReceived, noop] = fromContinuousCallback(trigger => {
  client.on('data', trigger)
});
```

WebSocket:
```javascript
const [messageReceived, unsubscribe] = fromContinuousCallback(trigger => {
  socket.on('event', trigger);
  return () => socket.off('event', trigger);
});
```

DOM element:
```javascript
const buttonClicked = fromEventTarget(button, "click");

const [buttonClicked, unsubscribe] = fromContinuousCallback(trigger => {
  const button = document.queryElement("#button");

  button.addEventListener('click', trigger);
  return () => button.removeEventListener('click', trigger);
});
```

setTimeout:
```typescript
const [anotherOneSecond, stopTimer] = fromContinuousCallback<number>(trigger => {
  let i = 0
  const id = setTimeout(() => trigger(i++), 1000)
  return () => clearTimeout(id)
});
```

etc...

**Comment by doasync** [comment]
Shouldn't it be called `fromEventSource` then?

**Comment by zerobias** [comment]
Usually, we add features that have predictable, unambiguous semantics and are difficult to implement otherwise. There are too many differences between api of libraries with this method

**Comment by AlexandrHoroshih** [comment]
I suggest to revive this idea in a slightly different way:

There is a common pattern to connect external event source to effector with scope support:
```ts
const connectSocketFx = attach({
 source: $socket,
 effect(socket) {
   socket.on("message", scopeBind(messageReceived))
 }
})
```
More details and problems with this solution here: https://github.com/effector/effector/issues/666

I suggest to add new api to cover all cases with external observables - with Fork API support
```ts
function fromReactiveSource<Payload, S>(config: {
 source?: Store<S>,
 setupOn: Clock<any>, // it is essential for Fork API support to have explicit unit to setup subscription in correct scope
 setup: (scopedTrigger: (p: Payload) => Payload, sourceData?: S) => void;
}): Event<Payload>

```
with usage like this:
```ts
const $history = createStore(createBrowserHistory())

const historyUpdated = fromReactiveSource<HistoryUpdate>({
 source: $history,
 setupOn: [appStarted, ...],
 setup: (scopedTrigger, history) => {
   history.listen(scopedTrigger)
 }
})
```
I think, it will allow to cover all cases with external subscriptions + scope support

Also, i think, it might be useful to have an option to provide destruction function and trigger:
```ts
const $history = createStore(createBrowserHistory())

const historyUpdated = fromReactiveSource<HistoryUpdate>({
 source: $history,
 setupOn: [appStarted, ...],
 setup: (scopedTrigger, history) => {
   const unlisten = history.listen(scopedTrigger);

   return () => unlisten()
 },
 destroyOn: appUnmounted
})
```
This will allow to avoid memory leaks during SSR usage, where user would need to create separate subscription for every scope (i.e. per request) and also destroy this subscription, when request is handled and scope should be thrown away

**Reply by igorkamyshev** [reply]
Sounds like a good method for `patronum` 🤔

**Reply by AlexandrHoroshih** [reply]
I think it is better to have something like that as a built-in api, since it is a more technical method 🤔

**Reply by zerobias** [reply]
Perhaps we should also add target support right away?

```ts

fromReactiveSource<HistoryUpdate>({
 source: $history,
 target: historyUpdated,
 setupOn: [appStarted],
 destroyOn: appUnmounted,
 setup(scopedTrigger, history) {
   const unlisten = history.listen(scopedTrigger)
   return () => unlisten()
 },
})
```

It helps to describe applications in a [cycle-free approach](https://github.com/effector/effector/issues/668#issuecomment-1118248363) right from the beginning
